runtime/runtime/src/lib.rs

```diff
@@ -1687,67 +1689,156 @@ impl Runtime {
+                .cmp(&(
+                    tx_b.transaction.signer_id(),
+                    tx_b.transaction.public_key(),
+                    tx_b.transaction.nonce(),
+                ))
+        });
+
+        let groups: Vec<(
+            (near_primitives::types::AccountId, near_crypto::PublicKey),
+            Vec<(SignedTransaction, TransactionCost)>,
+        )> = valid_txs
+            .into_iter()
+            .group_by(|(st, _)| {
+                (st.transaction.signer_id().clone(), st.transaction.public_key().clone())
+            })
+            .into_iter()
+            .map(|(key, group)| (key, group.collect()))
+            .collect();
+
+        #[derive(Debug)]
+        struct GroupTask {
+            ephemeral: EphemeralAccount,
+            group: Vec<(SignedTransaction, TransactionCost)>,
+        }
+
+        let mut tasks = Vec::new();
+        for ((signer_id, public_key), group_list) in groups {
+            let ephemeral = match EphemeralAccount::new(state_update, signer_id, public_key) {
+                Ok(e) => e,
+                Err(err) => return Err(err),
             };
+            tasks.push(GroupTask { ephemeral, group: group_list });
+        }
+
+        #[derive(Debug)]
+        struct GroupResult {
+            ephemeral: EphemeralAccount,
+            verified: Vec<(SignedTransaction, VerificationResult)>,
+        }
+
+        let group_outcomes: Vec<Result<GroupResult, InvalidTxError>> = tasks
+            .into_par_iter()
+            .map(|task| {
+                match verifier::verify_ephemeral_group(
+                    task.ephemeral,
+                    &task.group,
+                    Some(apply_state.block_height),
+                    apply_state.current_protocol_version,
+                    &apply_state.config,
+                ) {
+                    Ok((ephemeral, verified)) => Ok(GroupResult { ephemeral, verified }),
+                    Err(e) => Err(e),
+                }
+            })
+            .collect();
+
+        let mut ephemeral_accounts = Vec::new();
+        let mut all_verified = Vec::new();
+
+        for outcome in group_outcomes {
+            match outcome {
+                Ok(o) => {
+                    ephemeral_accounts.push(o.ephemeral);
+                    all_verified.extend(o.verified);
+                }
+                Err(e) => {
+                    Self::handle_invalid_transaction(
+                        e,
+                        &near_primitives::hash::CryptoHash::default(),
+                        processing_state.protocol_version,
+                        "ephemeral group checks error",
+                    )?;
+                }
+            }
+        }
+
+        for ephemeral in ephemeral_accounts {
+            ephemeral.commit(state_update);
+        }
 
-            let tx_result = self.process_transaction(
+        for (st, vr) in all_verified {
```

Comment:
I am not preserving original order at the momemnt, this might be a problem, will need to store original indices and sort one more time